File: python/bottle_song.py
MAP = [
    'no', 'One', 'Two', 'Three', 'Four', 'Five', 'Six', 'Seven', 'Eight',
    'Nine', 'Ten'
]


def plural(n):
    if n == 1:
        return ''
    return 's'


def step(n):
    return [
        f"{MAP[n]} green bottle{plural(n)} hanging on the wall,",
        f"{MAP[n]} green bottle{plural(n)} hanging on the wall,",
        "And if one green bottle should accidentally fall,",
        f"There'll be {MAP[n-1].lower()} green bottle{plural(n-1)} hanging on the wall."
    ]


def recite(start, take=1):
    res = []
    for i in range(take):
        lst = step(start)
        res.extend(lst)
        res.append("")
        start -= 1

    return res[:-1]

File: elixir/bottle_song.ex
defmodule BottleSong do
  @moduledoc """
  Handles lyrics of the popular children song: Ten Green Bottles
  """

  @dict %{0 => "no", 1 => "one", 2 => "two", 3 => "three", 4 => "four", 5 => "five", 6 => "six", 7 => "seven", 8 => "eight", 9 => "nine", 10 => "ten"}

  defp plural(1), do: ""
  defp plural(_), do: "s"

  @spec recite(pos_integer, pos_integer) :: String.t()
  def recite(start_bottle, take_down) do
    recite(start_bottle, take_down, [])
  end
  defp recite(_, 0, acc) do
    acc
    |> Enum.reverse()
    |> Enum.join("\n\n")
  end
  defp recite(start_bottle, take_down, acc) do
    s = """
    #{String.capitalize(@dict[start_bottle])} green bottle#{plural(start_bottle)} hanging on the wall,
    #{String.capitalize(@dict[start_bottle])} green bottle#{plural(start_bottle)} hanging on the wall,
    And if one green bottle should accidentally fall,
    There'll be #{@dict[start_bottle - 1]} green bottle#{plural(start_bottle-1)} hanging on the wall.\
    """
    recite(start_bottle - 1, take_down - 1, [s | acc])
  end
end

